polyply/src/linalg_functions.py

```diff
@@ -163,7 +163,23 @@ def _radius_of_gyration(points):
     return RG
 radius_of_gyration = jit(_radius_of_gyration)
 
-def rotate_xyz(object_xyz, theta_x, theta_y, theta_z):
+def _matrix_multiplication(*args):
+    matrix_a = args[0]
+    for matrix_b in args[1:]:
+        rows, other_dim = matrix_a.shape
+        columns = matrix_b.shape[1]
+        new_matrix = np.zeros((rows, columns))
+        for i in range(0, rows):
+            for j in range(0, columns):
+                for k in range(0, other_dim):
+                    new_matrix[i,j] += matrix_a[i,k] * matrix_b[k,j]
+
+        matrix_a = new_matrix
+    return new_matrix
```

Comment:
It is on purpose this pure and simple pythonic multiplication, because it compiles the best to numba, which gives overall a good performance. Amongst others see: https://stackoverflow.com/questions/36526708/comparing-python-numpy-numba-and-c-for-matrix-multiplication).  In addition matmul gives deprecated warnings in some numpy versions and also fundamentally a numpy array is not a matrix as in vector calculus and matmul is more a dirty trick around it. Therefore has been quite a bit of testing and thought going into this module as it is, it works the best across the board.

Reply:
I'm not sure you can compare directly to the results of the SO post, since your JIT compilation is much simpler (i.e. not the size constraints).
 
> matmul is more a dirty trick around it

Could you elaborate?

The reason I don't like the current form is because it's much less clear what's going on/what the purpose of the code is. A docstring and or comments would go a long way to resolving that if you're insistent on keeping this form. I do like my suggested signature more than yours though.

Also, this really wrecks performance for the people who don't have numba.

Reply:
Oh, and put your performance considerations in a comment so you won't "optimize" this yourself in the future.

Reply:
Couple of things: 

- matmul cannot handle multiplication of ndarrays with column vectors. Test 1 and 3 of the test-suit fail. I cannot gurantee the shape of the data to adhere to matmul rule, which I must admit is strange to me. 
-  this brings me to dirty trick: In my opinion it should treat np.ndarrays as np.matrix (conceptually speaking; independent of shape) and thus be able to do all vector calculus multiplications. But it does only multiply arrays and throws quite a some dimension errors and other warnings around. I didn't log them all, but it was enough for me to look up an alternative implementation. It also cannot handle symbolic data and there are other issues in my opinion with that function. 
- I don't want to touch matrix multiplications and rotations again for the time being. It took forever to figure out an ok way to implement it for the needs I have and test cases I ran. 
- There are so many other things depending on numba that it doesn't really matter. If you don't have it, the slow-down in the RW processor will be much more sever than in the backmapping. So this is something I can live with. 

A doc-string and comment logging the essence of our discussion, is rather appropriate though.

Reply:
You could have a look at `np.atleast_2d()` at some point, but we can leave it as-is for now.

> Also np.matmul does a ton of checking of the input dimensions. Like the entire module is full of try and excepts, which in general slows down functions quite a bit. So even if not directly comparable to SO, with my simple testes the numba speed-up is still given.

True.

> A doc-string and comment logging the essence of our discussion, is rather appropriate though.

That should be enough to put this to rest :)